### PR TITLE
DATACOUCH-287 - Add CAS operations to RxJavaCouchbaseOperations.

### PR DESCRIPTION
**Pull request by alexsderkach**
Currently, there is only one way to save an entity - upsert.

**RxJavaCouchbaseOperations** is missing important feature of Couchbase 2.x API - cas updates.
The idea is to provide a few methods, which will apply a function to existing/non-existing input document, and perform safe replace/insert.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@alexsderkach Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-couchbase&pullRequestId=135)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-data-couchbase&pullRequestId=135) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@alexsderkach Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-couchbase&pullRequestId=135)!

File: src/main/java/org/springframework/data/couchbase/core/RxJavaCouchbaseOperations.java [hunk 1]
```diff
@@ -53,6 +56,14 @@
 
     <T>Observable<T> remove(Iterable<T> batchToRemove, PersistTo persistTo, ReplicateTo replicateTo);
 
+    <T>Observable<T> cas(String id, Class<T> entityClass, Function<Optional<T>, CasResult<T>> updateFunction);
```
Comment (subalakr) [inline src/main/java/org/springframework/data/couchbase/core/RxJavaCouchbaseOperations.java]: The operations missing are insert, upsert and replace. Cas, asyncCas.. doesn't conform to the client library naming conventions

File: src/main/java/org/springframework/data/couchbase/core/RxJavaCouchbaseTemplate.java [hunk 2]
```diff
@@ -102,6 +113,74 @@
...
+    }
+
+    @Override
+    public <T> Observable<T> asyncCas(String id, Class<T> entityClass, PersistTo persistTo, ReplicateTo replicateTo, Function<Optional<T>, Observable<CasResult<T>>> updateFunction) {
+        return client.get(id, RawJsonDocument.class)
+                .flatMap(doc -> update(doc, entityClass, persistTo, replicateTo, updateFunction))
+                .switchIfEmpty(insert(entityClass, persistTo, replicateTo, updateFunction))
+                .retryWhen(casRetryFunction())
+                .onErrorResumeNext(e -> Observable.error(e.getCause()))
+                .doOnError(e -> TemplateUtils.translateError(e));
```
Comment (subalakr) [inline src/main/java/org/springframework/data/couchbase/core/RxJavaCouchbaseTemplate.java]: I don't think we need to do this in the library. The cas retries can be chained in by the application itself.

File: src/main/java/org/springframework/data/couchbase/core/RxJavaCouchbaseTemplate.java [hunk 3]
```diff
@@ -102,6 +113,74 @@
...
+                .switchIfEmpty(Observable.just(entity));
+    }
+
+    private <T> Observable.Transformer<Optional<T>, T> doModification(
+            Class<T> entityClass,
+            Function<Optional<T>, Observable<CasResult<T>>> updateFunction,
+            Function<RawJsonDocument, Observable<RawJsonDocument>> modificationFunction) {
+        return observable -> observable
+                .flatMap(updateFunction::apply)
+                .filter(this::shouldUpdate)
```
Comment (subalakr) [inline src/main/java/org/springframework/data/couchbase/core/RxJavaCouchbaseTemplate.java]: I think bubbling up the CASMismatchException, .. other exceptions in the casRetryFunction should be good enough. Spring retry can be used to do auto-retries, we don't usually provide this internally with the sdk itself as it is easily handled on the application end.

**Comment by alexsderkach** [comment]
@subalakr Updated according to requested changes

**Pull request merged**